pydantic_ai_slim/pydantic_ai/ui/ag_ui/_adapter.py

```diff
@@ -193,3 +202,150 @@ def load_messages(cls, messages: Sequence[Message]) -> list[ModelMessage]:
                     )
 
         return builder.messages
+
+    @classmethod
+    def dump_messages(cls, messages: Sequence[ModelMessage]) -> list[Message]:
+        """Transform Pydantic AI messages into AG-UI messages.
+
+        Note: AG-UI message IDs are not preserved from load_messages().
+
+        Args:
+            messages: Sequence of Pydantic AI [`ModelMessage`][pydantic_ai.messages.ModelMessage] objects.
+
+        Returns:
+            List of AG-UI protocol messages.
+        """
+        ag_ui_messages: list[Message] = []
+        message_id_counter = 1
+
+        def get_next_id() -> str:
+            nonlocal message_id_counter
+            result = f'msg_{message_id_counter}'
+            message_id_counter += 1
+            return result
+
+        for model_msg in messages:
+            if isinstance(model_msg, ModelRequest):
+                cls._convert_request_parts(model_msg.parts, ag_ui_messages, get_next_id)
+
+            elif isinstance(model_msg, ModelResponse):
+                cls._convert_response_parts(model_msg.parts, ag_ui_messages, get_next_id)
+
+        return ag_ui_messages
+
+    @staticmethod
+    def _convert_request_parts(
+        parts: Sequence[ModelRequestPart],
+        ag_ui_messages: list[Message],
+        get_next_id: Callable[[], str],
+    ) -> None:
+        """Convert ModelRequest parts to AG-UI messages."""
+        for part in parts:
+            msg_id = get_next_id()
+
+            if isinstance(part, SystemPromptPart):
+                ag_ui_messages.append(SystemMessage(id=msg_id, content=part.content))
+
+            elif isinstance(part, UserPromptPart):
+                content = part.content if isinstance(part.content, str) else str(part.content)
+                ag_ui_messages.append(UserMessage(id=msg_id, content=content))
+
+            elif isinstance(part, ToolReturnPart):
+                ag_ui_messages.append(
+                    ToolMessage(
+                        id=msg_id,
+                        content=AGUIAdapter._serialize_content(part.content),
+                        tool_call_id=part.tool_call_id,
+                    )
+                )
+
+    @staticmethod
+    def _convert_response_parts(
+        parts: Sequence[ModelResponsePart],
+        ag_ui_messages: list[Message],
+        get_next_id: Callable[[], str],
+    ) -> None:
+        """Convert ModelResponse parts to AG-UI messages."""
+
+        # Group consecutive assistant parts (text, tool calls) together
+        def is_assistant_part(part: ModelResponsePart) -> bool:
```

Comment:
Isn't everything in `ModelResponsePart` an assistant part?

Edit: Ah this is specifically because AG-UI doesn't understand builtin tool returns that come from the assistant. Let's make that explicit.